runtime/neumann/src/fee_handler.rs

```diff
@@ -0,0 +1,163 @@
+	) -> Result<Self::LiquidityInfo, TransactionValidityError> {
+		if fee.is_zero() {
+			return Ok(None)
+		}
+
+		// cross it
+		let call_information = FCP::fee_information(call.clone());
+
+		// TODO: when no info or we don't allow that token throw error
+
+		if call_information.token_id == NATIVE_TOKEN_ID {
+			let withdraw_reason = if tip.is_zero() {
+				WithdrawReasons::TRANSACTION_PAYMENT
+			} else {
+				WithdrawReasons::TRANSACTION_PAYMENT | WithdrawReasons::TIP
+			};
+
+			match C::withdraw(who, fee, withdraw_reason, ExistenceRequirement::KeepAlive) {
+				Ok(imbalance) => Ok(Some(imbalance)),
+				Err(_) => Err(InvalidTransaction::Payment.into()),
+			}
+		} else {
+			let currency_id = call_information.token_id.into();
+			// orml_tokens doesn't provide withdraw that allows setting existence so prevent
+			// withdraw if they don't have enough to avoid reaping
+			MC::ensure_can_withdraw(
+				currency_id,
+				who,
+				fee.saturating_add(MC::minimum_balance(currency_id).into()).into(),
+			)
+			.map_err(|_| TransactionValidityError::Invalid(InvalidTransaction::Payment))?;
+
+			MC::withdraw(currency_id, who, fee.into())
+				.map_err(|_| TransactionValidityError::Invalid(InvalidTransaction::Payment))?;
+
+			MC::deposit(currency_id, &TA::get(), fee.into()) // treasury account
+				.map_err(|_| TransactionValidityError::Invalid(InvalidTransaction::Payment))?;
+
+			// TODO: Fire event for deposit
+
+			// We dealt with imbalance here so don't let `correct_and_deposit_fee` do it
+			Ok(None)
+		}
+	}
+
+	/// Hand the fee and the tip over to the `[OnUnbalanced]` implementation.
+	/// Since the predicted fee might have been too high, parts of the fee may
+	/// be refunded.
+	///
+	/// Note: The `corrected_fee` already includes the `tip`.
+	fn correct_and_deposit_fee(
+		who: &T::AccountId,
+		_dispatch_info: &DispatchInfoOf<T::Call>,
+		_post_info: &PostDispatchInfoOf<T::Call>,
+		corrected_fee: Self::Balance,
+		tip: Self::Balance,
+		already_withdrawn: Self::LiquidityInfo,
+	) -> Result<(), TransactionValidityError> {
+		if let Some(paid) = already_withdrawn {
+			// Calculate how much refund we should return
+			let refund_amount = paid.peek().saturating_sub(corrected_fee);
+			// refund to the the account that paid the fees. If this fails, the
+			// account might have dropped below the existential balance. In
+			// that case we don't refund anything.
+			let refund_imbalance = C::deposit_into_existing(who, refund_amount)
+				.unwrap_or_else(|_| C::PositiveImbalance::zero());
+			// merge the imbalance caused by paying the fees and refunding parts of it again.
+			let adjusted_paid = paid
+				.offset(refund_imbalance)
+				.same()
+				.map_err(|_| TransactionValidityError::Invalid(InvalidTransaction::Payment))?;
+			// Call someone else to handle the imbalance (fee and tip separately)
+			let (tip, fee) = adjusted_paid.split(tip);
+			OU::on_unbalanceds(Some(fee).into_iter().chain(Some(tip)));
+		}
+		Ok(())
```

Comment:
Matches `pallet_transaction_payment` `CurrencyAdapter`.

Reply:
Does it make sense to forward to the transaction payment pallet here rather than copying the code?
```
pallet_transaction_payment::CurrencyAdapter<Balances, DealWithInclusionFees<Runtime>>::correct_and_deposit_fee(...)
```